File: sdk/dotnet/src/Cordum.Cap/Validate.cs
using Cordum.Agent.V1;

namespace Cordum.Cap;

/// <summary>
/// Input validation matching the CAP guard SDK rules.
/// </summary>
public static class Validate
{
    /// <summary>
    /// Validate a JobRequest. Throws ValidationException on invalid input.
    /// </summary>
    public static void ValidateJobRequest(JobRequest req)
    {
        if (string.IsNullOrEmpty(req.JobId))
            throw new ValidationException("job_id", "must not be empty");
        if (string.IsNullOrEmpty(req.Topic))
            throw new ValidationException("topic", "must not be empty");
        if (req.StepIndex < 0)
            throw new ValidationException("step_index", "must be >= 0");
        if (req.Budget != null)
        {
            if (req.Budget.MaxInputTokens < 0)
                throw new ValidationException("budget.max_input_tokens", "must be >= 0");
            if (req.Budget.MaxOutputTokens < 0)
                throw new ValidationException("budget.max_output_tokens", "must be >= 0");
            if (req.Budget.MaxTotalTokens < 0)
                throw new ValidationException("budget.max_total_tokens", "must be >= 0");
            if (req.Budget.DeadlineMs < 0)
                throw new ValidationException("budget.deadline_ms", "must be >= 0");
        }
    }

    /// <summary>
    /// Validate a JobResult. Throws ValidationException on invalid input.
    /// </summary>
    public static void ValidateJobResult(JobResult res)
    {
        if (string.IsNullOrEmpty(res.JobId))
            throw new ValidationException("job_id", "must not be empty");
        if (res.Status == JobStatus.Unspecified)
            throw new ValidationException("status", "must not be UNSPECIFIED");
        if (string.IsNullOrEmpty(res.WorkerId))
            throw new ValidationException("worker_id", "must not be empty");
        if (res.ExecutionMs < 0)
            throw new ValidationException("execution_ms", "must be >= 0");
    }

    /// <summary>
    /// Validate a BusPacket envelope. Throws ValidationException on invalid input.
    /// </summary>
    public static void ValidateBusPacket(BusPacket pkt)
    {
        if (string.IsNullOrEmpty(pkt.TraceId))
            throw new ValidationException("trace_id", "must not be empty");
        if (string.IsNullOrEmpty(pkt.SenderId))
            throw new ValidationException("sender_id", "must not be empty");
        if (pkt.ProtocolVersion <= 0)
            throw new ValidationException("protocol_version", "must be > 0");
        if (pkt.CreatedAt == null)
            throw new ValidationException("created_at", "must not be null");
        if (pkt.PayloadCase == BusPacket.PayloadOneofCase.None)
            throw new ValidationException("payload", "must not be empty");

        switch (pkt.PayloadCase)
        {
            case BusPacket.PayloadOneofCase.JobRequest:
                ValidateJobRequest(pkt.JobRequest);
                break;
            case BusPacket.PayloadOneofCase.JobResult:
                ValidateJobResult(pkt.JobResult);
                break;
        }
    }
}

File: sdk/java/src/main/java/io/cordum/cap/Validate.java
package io.cordum.cap;

import io.cordum.cap.agent.v1.BusPacket;
import io.cordum.cap.agent.v1.JobRequest;
import io.cordum.cap.agent.v1.JobResult;
import io.cordum.cap.agent.v1.JobStatus;
import io.cordum.cap.agent.v1.JobPriority;
import io.cordum.cap.agent.v1.Budget;

public final class Validate {

    private Validate() {}

    public static void validateJobRequest(JobRequest req) {
        if (req == null) {
            throw new Errors.ValidationException("job_request", "must not be null");
        }
        if (req.getJobId().isEmpty()) {
            throw new Errors.ValidationException("job_id", "must not be empty");
        }
        if (req.getTopic().isEmpty()) {
            throw new Errors.ValidationException("topic", "must not be empty");
        }
        if (req.getPriorityValue() < 0
                || req.getPriorityValue() > JobPriority.JOB_PRIORITY_CRITICAL_VALUE) {
            throw new Errors.ValidationException("priority", "out of valid range");
        }
        if (req.getStepIndex() < 0) {
            throw new Errors.ValidationException("step_index", "must be >= 0");
        }
        if (req.hasBudget()) {
            Budget b = req.getBudget();
            if (b.getMaxInputTokens() < 0) {
                throw new Errors.ValidationException("budget.max_input_tokens", "must be >= 0");
            }
            if (b.getMaxOutputTokens() < 0) {
                throw new Errors.ValidationException("budget.max_output_tokens", "must be >= 0");
            }
            if (b.getMaxTotalTokens() < 0) {
                throw new Errors.ValidationException("budget.max_total_tokens", "must be >= 0");
            }
            if (b.getDeadlineMs() < 0) {
                throw new Errors.ValidationException("budget.deadline_ms", "must be >= 0");
            }
        }
    }

    public static void validateJobResult(JobResult res) {
        if (res == null) {
            throw new Errors.ValidationException("job_result", "must not be null");
        }
        if (res.getJobId().isEmpty()) {
            throw new Errors.ValidationException("job_id", "must not be empty");
        }
        if (res.getStatus() == JobStatus.JOB_STATUS_UNSPECIFIED) {
            throw new Errors.ValidationException("status", "must not be UNSPECIFIED");
        }
        if (res.getWorkerId().isEmpty()) {
            throw new Errors.ValidationException("worker_id", "must not be empty");
        }
        if (res.getExecutionMs() < 0) {
            throw new Errors.ValidationException("execution_ms", "must be >= 0");
        }
    }

    public static void validateBusPacket(BusPacket pkt) {
        if (pkt == null) {
            throw new Errors.ValidationException("bus_packet", "must not be null");
        }
        if (pkt.getTraceId().isEmpty()) {
            throw new Errors.ValidationException("trace_id", "must not be empty");
        }
        if (pkt.getSenderId().isEmpty()) {
            throw new Errors.ValidationException("sender_id", "must not be empty");
        }
        if (pkt.getProtocolVersion() <= 0) {
            throw new Errors.ValidationException("protocol_version", "must be > 0");
        }
        if (!pkt.hasCreatedAt()) {
            throw new Errors.ValidationException("created_at", "must not be null");
        }
        if (pkt.getPayloadCase() == BusPacket.PayloadCase.PAYLOAD_NOT_SET) {
            throw new Errors.ValidationException("payload", "must not be empty");
        }
        if (pkt.getPayloadCase() == BusPacket.PayloadCase.JOB_REQUEST) {
            validateJobRequest(pkt.getJobRequest());
        }
        if (pkt.getPayloadCase() == BusPacket.PayloadCase.JOB_RESULT) {
            validateJobResult(pkt.getJobResult());
        }
    }
}
